### PR TITLE
cola_var update

### PR DESCRIPTION
**Pull request by PterosDiacos**

### CODE REVIEW

**Comment by sleepinyourhat** [comment]
Should we merge this or #547? They look extremely similar.

**Comment by PterosDiacos** [comment]
> Should we merge this or #547? They look extremely similar.

The repository for #547 has been removed. Still debugging the set-up for sequential training for MNLI

**Comment by sleepinyourhat** [comment]
Hrm—switch back to 547 if you can find an easy way to: It's definitely still there and mergeable. In any case, though, let me know when this needs another look.

**Comment by sleepinyourhat** [comment]
Ah, nvm. I see the issue. Closing #547.

**Comment by PterosDiacos** [comment]
@sleepinyourhat : I tried @Sheng-Fu and @zphang 's recipe but have not made sequential training work. One thing is that `main.py` does not allow `target_tasks` set to be `none`. But having the latter set to some random task, I ran into the error posted on Slack

**Comment by sleepinyourhat** [comment]
Ready for another look?

File: config/spring19_seminar/bow_glove.conf [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+write_preds = "val,test"  // 0 for none, or comma-separated splits in {"train", "val", "test"} 
+                          // for which predictions are written to disk during do_full_eval
+
+
+// Pretraining tasks
+load_model = 0  // If true, restore from checkpoint when starting do_pretrain
+
+
+// Models
+// Model, bow + GloVe
```
Comment (sheng-fu) [inline config/spring19_seminar/bow_glove.conf]: For some reason I also need bert_model_name = "" to make it work

Comment (sheng-fu) [reply config/spring19_seminar/bow_glove.conf]: I also needed "allow_missing_task_map = 1"

**Comment by sheng-fu** [comment]
Clarification question: in the sbatch scripts, mnli and ccg are pre-training tasks, while cola and all other npi settings are target tasks. Crucially, all of these would fine-tune BERT. Is this the setting we want? 

If so, then we wouldn't have the problem of two-stage pretraining, since we would just be doing pretraining and target-task training (and it makes things easier I think?).

One minor adjustment is that instead of fine-tuning BERT on mnli and cgg 11 times each, we can to do each just once, and then for each of the target-task settings, we load the pretrained checkpoints.

File: scripts/spring19_seminar/cola_npi_x.sh [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+# These 9 different environments are to be found at https://github.com/alexwarstadt/data_generation/tree/master/outputs/npi/environments/splits
+# Finetuning on one of the NPI task, we target all NPI tasks;
+# Finetuning on eight of the NPI tasks, we target the remaining one.
+
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola, target_tasks = \"cola,cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola\", do_pretrain = 1" 
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola_npi_adv\", do_pretrain = 1"
```
Comment (sheng-fu) [inline scripts/spring19_seminar/cola_npi_x.sh]: This will make all the target-tasks fine-tune BERT! I think we probably don't want to do this, right? I 
`python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"cola_npi_adv\", do_pretrain = 1"

python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"cola_npi_adv,cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", use_classifier = cola_npi_adv" 
`

File: scripts/spring19_seminar/cola_npi_x.sh [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola, target_tasks = \"cola,cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola\", do_pretrain = 1" 
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola_npi_adv\", do_pretrain = 1" 
+
+python main.py --config_file config/spring19_seminar/bert.conf \
+    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_cond, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola_npi_cond\", do_pretrain = 1"
```
Comment (sheng-fu) [inline scripts/spring19_seminar/cola_npi_x.sh]: We should probably separate pretraining and target-task training into two steps, since the current setting would make all the target_tasks fine-tune BERT, if I'm not mistaken.

Separation may be done like this:
python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"\", pretrain_tasks = \"cola_npi_adv\", do_pretrain = 1" 

python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", transfer_paradigm = \"finetune\""

File: config/spring19_seminar/bow_glove.conf (hunk 1 above)
Comment (PterosDiacos) [reply config/spring19_seminar/bow_glove.conf]: `bert_model_name = ""` is in `defaults.conf`

File: scripts/spring19_seminar/cola_npi_x.sh (hunk 3 above)
Comment (PterosDiacos) [reply scripts/spring19_seminar/cola_npi_x.sh]: Currently all "fine-tuning tasks" have been treated as "pretraining tasks".


```python main.py --config_file config/spring19_seminar/bert.conf \
    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_cond, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", pretrain_tasks = \"cola_npi_cond\", do_pretrain = 1" ```

The intention of the above cmd is like

(1) train (plain) BERT on cola_npi_cond;
(2) test BERT of (1) on cola_npi_sup;
(3) test BERT of (1) on cola_npi_quessmp;
...
(n) test BERT of (1) on cola_npi_adv.

Since do_target_task_training has been turned off, it might not be a problem.
I remembered @HaokunLiu  mentioned in Slack that even if we do_target_task_training, the execution is still in a "distributive" manner: for each cola_npi_x  BERT of (1) is finetuned on cola_npi_x and eval on cola_npi_x

Can @HaokunLiu  @sleepinyourhat  confirm this?

Comment (PterosDiacos) [reply scripts/spring19_seminar/cola_npi_x.sh]: @sheng-fu :  maybe I misunderstood something. But if do_target_task_training first fine-tunes the model on *all* target tasks before evaluate one after another, the following separation might still have the same problem

> `Separation may be done like this:
python main.py --config_file config/spring19_seminar/bert.conf 
--overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = "", pretrain_tasks = "cola_npi_adv", do_pretrain = 1"

python main.py --config_file config/spring19_seminar/bert.conf 
--overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_adv, target_tasks = "cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv", transfer_paradigm = "frozen""`

Comment (sheng-fu) [reply scripts/spring19_seminar/cola_npi_x.sh]: Oh I see, so your commands are just pre-training commands. The target-tasks do not really matter then? So I guess my point would be, in evaluation, we should make sure that transfer_paradigm is frozen.

    --overrides "exp_name = npi_bert, run_name = run_bertnone_cola_npi_cond, target_tasks = \"cola_npi_sup,cola_npi_quessmp,cola_npi_ques,cola_npi_qnt,cola_npi_only,cola_npi_negsent,cola_npi_negdet,cola_npi_cond,cola_npi_adv\", do_target_task_training = 1,  transfer_paradigm = \"frozen\""

----

By the way, all this relies on cgg/mnli and cola/npi tasks are both pretraining tasks and can maybe be pretrained one after another, right?

Comment (sheng-fu) [reply scripts/spring19_seminar/cola_npi_x.sh]: Oh sorry I missed the part about doing full eval without doing target_task_training. Then you're right, what I raised is not a problem.

This still relies on bert+ccg and bert+mnli being able to undergo further pretraining, right? I tried today and couldn't get that to work. Should we just throw ccg and mnli as part of the pretraining tasks together with the cola/npi tasks?

**Comment by sheng-fu** [comment]
A very minor suggestion: I think it's better to put our NPI data in a separate folder from the CoLA data. I also suggest just using the folder names and folder organization in the data_generation repo, so that we don't have to change the names of the folder in case of last-minute data revision. 

But this doesn't affect how we run experiments, so feel free to ignore it.

File: src/tasks/tasks.py [hunk 4]
```diff
@@ -415,6 +415,25 @@ def load_data(self, path, max_seq_len):
...
+@register_task('all_cola_npi', rel_path='CoLA/all_npi')
+@register_task('cola_npi_sup', rel_path='CoLA/npi_sup')
+@register_task('cola_npi_quessmp', rel_path='CoLA/npi_quessmp')
+@register_task('cola_npi_ques', rel_path='CoLA/npi_ques')
+@register_task('cola_npi_qnt', rel_path='CoLA/npi_qnt')
+@register_task('cola_npi_only', rel_path='CoLA/npi_only')
+@register_task('cola_npi_negsent', rel_path='CoLA/npi_negsent')
+@register_task('cola_npi_negdet', rel_path='CoLA/npi_negdet')
+@register_task('cola_npi_cond', rel_path='CoLA/npi_cond')
+@register_task('cola_npi_adv', rel_path='CoLA/npi_adv')
```
Comment (sheng-fu) [inline src/tasks/tasks.py]: I suggest moving these to a different task that loads test_full.tsv for the test set. This would make test-set evaluation so much easier.

e.g.,
te_data = load_tsv(self._tokenizer_name, os.path.join(path, 'test_full.tsv'), max_seq_len,
                           s1_idx=3, s2_idx=None, label_idx=1)

File: src/tasks/tasks.py [hunk 5]
```diff
@@ -434,6 +434,45 @@ def load_data(self, path, max_seq_len):
...
+        tr_data = load_tsv(self._tokenizer_name, os.path.join(path, "train.tsv"), max_seq_len,
+                           s1_idx=3, s2_idx=None, label_idx=1)
+        val_data = load_tsv(self._tokenizer_name, os.path.join(path, "dev.tsv"), max_seq_len,
+                            s1_idx=3, s2_idx=None, label_idx=1)
+        te_data = load_tsv(self._tokenizer_name, os.path.join(path, 'test_full.tsv'), max_seq_len,
+                            s1_idx=3, s2_idx=None, label_idx=1)
+        self.train_data_text = tr_data
+        self.val_data_text = val_data
+        self.test_data_text = te_data
+        log.info("\tFinished loading CoLA.")
```
Comment (sleepinyourhat) [inline src/tasks/tasks.py]: Rename

File: src/tasks/tasks.py [hunk 6]
```diff
@@ -434,6 +434,45 @@ def load_data(self, path, max_seq_len):
 @register_task('cola_npi_negdet', rel_path='CoLA/npi_negdet')
 @register_task('cola_npi_cond', rel_path='CoLA/npi_cond')
 @register_task('cola_npi_adv', rel_path='CoLA/npi_adv')
+class NPITask(SingleClassificationTask):
+    '''Class for NPI-related task; same with Warstdadt acceptability task but outputs labels for test-set'''
```
Comment (sleepinyourhat) [inline src/tasks/tasks.py]: => CoLA?

Comment (sleepinyourhat) [reply src/tasks/tasks.py]: Especially since we're using the CoLA directory and cola_ in the task names.

Comment (sheng-fu) [reply src/tasks/tasks.py]: This is actually another point that I've raised a couple of times... I think it's kind of confusing to put these in the CoLA folder and also have "cola" in the task name... I have suggested put them in another folder (maybe called "NPI") and also just use the folder structure in the data_generation folder. 

But if we're sticking to the current folder/task arrangement, maybe I can call this something like CoLANPITask?

**Comment by sleepinyourhat** [comment]
Ready?

(Also, there are merge conflicts you'll have to fix before merging.)

**Comment by W4ngatang** [comment]
Is this still alive / something we're hoping to merge in soon? Could you give a description for what's going on in this PR (or if Sam is overseeing it and knows, fine without).

**Comment by sleepinyourhat** [comment]
Please finish this soon. Before the seminar paper comes out, we'll need clear documentation for how to reproduce the experiments, and that'll be hard to put together if we wait.

**Comment by PterosDiacos** [comment]
> Please finish this soon. Before the seminar paper comes out, we'll need clear documentation for how to reproduce the experiments, and that'll be hard to put together if we wait.

Done:)

**Comment by sleepinyourhat** [comment]
Thanks! It's not letting us merge this because of a style error—I'm guessing it's the new line you added to evaluate.py.

**Comment by PterosDiacos** [comment]
hmm there should be no commitments to `evaluate.py`
